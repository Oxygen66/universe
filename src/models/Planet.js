import database from '../../database';

class Planet {
  #db = database;

  async getAll() {
    return this.#db('planet');
  }

  async getBySpaceCenterId(id) {
    return this.#db('planet')
      .select('planet.*')
      .innerJoin(
        'space_center',
        'space_center.planet_id',
        'planet.id',
      )
      .where('space_center.id', id)
      .first();
  }
}

export default Planet;
