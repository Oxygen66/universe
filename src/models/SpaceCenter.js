import database from '../../database';

class SpaceCenter {
  #db = database;

  async getByPlanetIdAndLimit(planetId, limit) {
    return this.#db('space_center')
      .where('planet_id', planetId)
      .limit(limit);
  }

  async getById(id) {
    return this.#db('space_center')
      .where({ id })
      .first();
  }

  async getByUid(uid) {
    return this.#db('space_center')
      .where({ uid })
      .first();
  }

  async getAllPaginated(page, pageSize) {
    return this.#db('space_center')
      .limit(pageSize)
      .offset((pageSize - 1) * page);
  }

  async countAll() {
    return this.#db('space_center').count({ total: 'id' });
  }
}

export default SpaceCenter;
