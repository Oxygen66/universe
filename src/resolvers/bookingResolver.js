import { UserInputError } from 'apollo-server-koa';

const bookingResolver = {
  Mutation: {
    bookFlight: async (
      _,
      { bookingInfo: { seatCount, flightId, email } },
      { dataSources: { bookingAPI, flightAPI } },
    ) => {
      if (seatCount < 1) {
        throw new UserInputError('seatCount should be greater than 0', {
          invalidArgs: 'seatCount',
        });
      }

      const flight = await flightAPI.getById(flightId);

      const bookings = await bookingAPI.getAllByFlightId(flightId);
      const remainingSeats = [...bookings]
        .map((booking) => booking.seat_count)
        .reduce((previousValue, currentValue) => previousValue - currentValue, flight.seat_count);

      if ((remainingSeats - seatCount) < 0) {
        throw new UserInputError('seatCount must be less than the number of remaining seats', {
          invalidArgs: 'seatCount',
        });
      }

      const [bookingResult] = await bookingAPI.createBooking(seatCount, flightId, email);
      return bookingResult;
    },
  },
  Query: {
    booking: (_, { id }, { dataSources: { bookingAPI } }) => bookingAPI.getById(id),
    bookings: async (parent, { email, page, pageSize }, { dataSources: { bookingAPI } }) => {
      if (page < 1) {
        throw new UserInputError('Page must be greater than 1', {
          invalidArgs: 'page',
        });
      }

      if (pageSize < 1 || pageSize > 100) {
        throw new UserInputError('PageSize must be between 1 and 100', {
          invalidArgs: 'page',
        });
      }

      return {
        bookings: email
          ? await bookingAPI.getAllByEmailPaginated(email, page, pageSize)
          : await bookingAPI.getAllPaginated(page, pageSize),
        page,
        pageSize,
        email,
      };
    },
  },
  Booking: {
    id: (parent) => parent.id,
    seatCount: (parent) => parent.seat_count,
    email: (parent) => parent.email,
    flight: (parent, args, { dataSources: { flightAPI } }) => flightAPI.getById(parent.flight_id),
  },
  BookingsResult: {
    pagination: async ({ page, pageSize, email }, args, { dataSources: { bookingAPI } }) => {
      const [{ total }] = await (email
        ? bookingAPI.countByEmail(email)
        : bookingAPI.count()
      );
      return {
        total,
        pageSize,
        page,
      };
    },
    nodes: ({ bookings }) => bookings,
  },
};

export default bookingResolver;
