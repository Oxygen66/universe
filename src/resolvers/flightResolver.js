import { UserInputError } from 'apollo-server-koa';

const flightResolver = {
  Query: {
    flight: (parent, { id }, { dataSources: { flightAPI } }) => flightAPI.getById(id),
    flights:
      async (
        parent,
        {
          page, pageSize, seatCount, from, to, departureDay,
        },
        { dataSources: { flightAPI } },
      ) => {
        if (page < 1) {
          throw new UserInputError('Page must be greater than 1', {
            invalidArgs: 'page',
          });
        }

        if (pageSize < 1 || pageSize > 100) {
          throw new UserInputError('PageSize must be between 1 and 100', {
            invalidArgs: 'page',
          });
        }

        return {
          flights: await flightAPI
            .getAllPaginated(page, pageSize, from, to, seatCount, departureDay),
          page,
          pageSize,
          from,
          to,
          seatCount,
          departureDay,
        };
      },
  },
  Mutation: {
    scheduleFlight:
      async (
        _,
        {
          flightInfo: {
            launchSiteId,
            landingSiteId,
            departureAt,
            seatCount,
          },
        },
        { dataSources: { flightAPI } },
      ) => {
        if (landingSiteId === launchSiteId) {
          throw new UserInputError('landingSiteId and landingSiteId cannot be the same', {
            invalidArgs: [
              'launchSiteId', 'landingSiteId',
            ],
          });
        }

        if (!departureAt) {
          throw new UserInputError('departureAt must be a valid ISO date', {
            invalidArgs: 'departureAt',
          });
        }

        const [flightResult] = await flightAPI
          .createFlight(launchSiteId, landingSiteId, departureAt, seatCount);
        return flightResult;
      },
  },
  Flight: {
    id: (parent) => parent.id,
    code: (parent) => parent.code,
    departureAt: (parent) => parent.departure_at,
    seatCount: (parent) => parent.seat_count,
    launchSite: (
      parent, args, { dataSources: { spaceCenterAPI } },
    ) => spaceCenterAPI.getById(parent.launching_site),
    landingSite: (
      parent, args, { dataSources: { spaceCenterAPI } },
    ) => spaceCenterAPI.getById(parent.landing_site),
    availableSeats: async (
      { id, seat_count: seatCount }, args, { dataSources: { bookingAPI } },
    ) => {
      const bookings = await bookingAPI.getAllByFlightId(id);
      return [...bookings]
        .map((booking) => booking.seat_count)
        .reduce((previousValue, currentValue) => previousValue - currentValue, seatCount);
    },
  },
  FlightsResult: {
    pagination: async (
      {
        page, pageSize, from, to, seatCount, departureDay,
      },
      args,
      { dataSources: { flightAPI } },
    ) => {
      const [{ total }] = await flightAPI.countAll(from, to, seatCount, departureDay);
      return {
        total,
        page,
        pageSize,
      };
    },
    nodes: (parent) => parent.flights,
  },
};

export default flightResolver;
