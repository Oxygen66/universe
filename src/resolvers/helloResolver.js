const helloResolvers = {
  Query: {
    hello: () => 'Hello Adrien',
  },
};

export default helloResolvers;
