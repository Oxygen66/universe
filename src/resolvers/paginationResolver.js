const paginationResolver = {
  Pagination: {
    total: (parent) => parent.total,
    page: (parent) => parent.page,
    pageSize: (parent) => parent.pageSize,
  },
};

export default paginationResolver;
