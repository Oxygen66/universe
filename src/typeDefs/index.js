import types from './types';
import query from './query';
import mutation from './mutation';
import inputs from './inputs';

const typeDefs = [
  ...types,
  ...inputs,
  query,
  mutation,
];

export default typeDefs;
