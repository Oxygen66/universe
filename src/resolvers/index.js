import { merge } from 'lodash';
import planetResolver from './planetResolver';
import spaceCenterResolver from './spaceCenterResolver';
import paginationResolver from './paginationResolver';
import flightResolver from './flightResolver';

const resolvers = merge(
  planetResolver,
  spaceCenterResolver,
  paginationResolver,
  flightResolver,
);

export default resolvers;
