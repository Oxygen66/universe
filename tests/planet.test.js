import { ApolloServer, gql } from 'apollo-server-koa';
import { createTestClient } from 'apollo-server-testing';
import Planet from '../src/models/Planet';
import SpaceCenter from '../src/models/SpaceCenter';
import typeDefs from '../src/typeDefs';
import resolvers from '../src/resolvers';

const planetAPI = new Planet();
const spaceCenterAPI = new SpaceCenter();

const GET_PLANETS = gql`
    query planets {
        planets {
            id
            name
            code
            spaceCenters {
                id
            }
        }
    }
`;

const GET_PLANETS_BY_LIMIT = gql`
    query planets($limit: Int!) {
        planets {
            id
            name
            code
            spaceCenters(limit: $limit) {
                id
            }
        }
    }
`;

const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: () => ({ planetAPI, spaceCenterAPI }),
});

describe('Queries', () => {
  beforeEach(() => {
    planetAPI.getAll = jest.fn(() => ([
      {
        id: 1,
        name: 'test1',
        code: 'CD1',
      },
      {
        id: 2,
        name: 'test2',
        code: 'CD',
      },
    ]));
    spaceCenterAPI.getByPlanetIdAndLimit = jest.fn((id, limit) => [...Array(limit)].map((_, i) => ({
      id: (id - 1) * limit + i + 1,
    })));
  });

  it('fetch list of planets', async () => {
    const { query } = createTestClient(server);
    const res = await query({
      query: GET_PLANETS,
    });
    res.data.planets.forEach((planet) => {
      expect(planet.spaceCenters).toHaveLength(5);
    });
    expect(res.errors).not.toBeTruthy();
    expect(res).toMatchSnapshot();
  });

  it('fetch list of planets with custom limit', async () => {
    const { query } = createTestClient(server);
    const limit = 7;
    const res = await query({
      query: GET_PLANETS_BY_LIMIT,
      variables: {
        limit,
      },
    });
    res.data.planets.forEach((planet) => {
      expect(planet.spaceCenters).toHaveLength(limit);
    });
    expect(res.errors).not.toBeTruthy();
    expect(res).toMatchSnapshot();
  });

  it('fetch list of planets with custom limit throw error (< 1)', async () => {
    const { query } = createTestClient(server);
    const res = await query({
      query: GET_PLANETS_BY_LIMIT,
      variables: {
        limit: 0,
      },
    });
    expect(res.errors).toBeTruthy();
    expect(res).toMatchSnapshot();
  });

  it('fetch list of planets with custom limit throw error (> 10)', async () => {
    const { query } = createTestClient(server);
    const res = await query({
      query: GET_PLANETS_BY_LIMIT,
      variables: {
        limit: 11,
      },
    });
    expect(res.errors).toBeTruthy();
    expect(res).toMatchSnapshot();
  });
});
