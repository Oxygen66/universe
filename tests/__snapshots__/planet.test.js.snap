// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries planets fetch list of planets 1`] = `
Object {
  "data": Object {
    "planets": Array [
      Object {
        "code": "CD1",
        "id": "1",
        "name": "test1",
        "spaceCenters": Array [
          Object {
            "id": "1",
          },
          Object {
            "id": "2",
          },
          Object {
            "id": "3",
          },
          Object {
            "id": "4",
          },
          Object {
            "id": "5",
          },
        ],
      },
      Object {
        "code": "CD",
        "id": "2",
        "name": "test2",
        "spaceCenters": Array [
          Object {
            "id": "6",
          },
          Object {
            "id": "7",
          },
          Object {
            "id": "8",
          },
          Object {
            "id": "9",
          },
          Object {
            "id": "10",
          },
        ],
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries planets fetch list of planets with custom limit 1`] = `
Object {
  "data": Object {
    "planets": Array [
      Object {
        "code": "CD1",
        "id": "1",
        "name": "test1",
        "spaceCenters": Array [
          Object {
            "id": "1",
          },
          Object {
            "id": "2",
          },
          Object {
            "id": "3",
          },
          Object {
            "id": "4",
          },
          Object {
            "id": "5",
          },
          Object {
            "id": "6",
          },
          Object {
            "id": "7",
          },
        ],
      },
      Object {
        "code": "CD",
        "id": "2",
        "name": "test2",
        "spaceCenters": Array [
          Object {
            "id": "8",
          },
          Object {
            "id": "9",
          },
          Object {
            "id": "10",
          },
          Object {
            "id": "11",
          },
          Object {
            "id": "12",
          },
          Object {
            "id": "13",
          },
          Object {
            "id": "14",
          },
        ],
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries planets fetch list of planets with custom limit throw error (< 1) 1`] = `
Object {
  "data": Object {
    "planets": Array [
      Object {
        "code": "CD1",
        "id": "1",
        "name": "test1",
        "spaceCenters": null,
      },
      Object {
        "code": "CD",
        "id": "2",
        "name": "test2",
        "spaceCenters": null,
      },
    ],
  },
  "errors": Array [
    [GraphQLError: Limit must be between 1 and 10],
    [GraphQLError: Limit must be between 1 and 10],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries planets fetch list of planets with custom limit throw error (> 10) 1`] = `
Object {
  "data": Object {
    "planets": Array [
      Object {
        "code": "CD1",
        "id": "1",
        "name": "test1",
        "spaceCenters": null,
      },
      Object {
        "code": "CD",
        "id": "2",
        "name": "test2",
        "spaceCenters": null,
      },
    ],
  },
  "errors": Array [
    [GraphQLError: Limit must be between 1 and 10],
    [GraphQLError: Limit must be between 1 and 10],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
